src/builtins/impls/time.rs

```diff
@@ -14,67 +14,147 @@
-pub fn date(x: serde_json::Value) -> Result<(u8, u8, u8)> {
-    bail!("not implemented");
+pub fn date(x: serde_json::Value) -> Result<(i32, u32, u32)> {
+    let date_time = extract_data_from_value(&x)?;
+    Ok((date_time.year(), date_time.month(), date_time.day()))
 }
 
 /// Returns the difference between two unix timestamps in nanoseconds (with
 /// optional timezone strings).
 #[tracing::instrument(name = "time.diff", err)]
+// todo:: need to implement
 pub fn diff(ns1: serde_json::Value, ns2: serde_json::Value) -> Result<(u8, u8, u8, u8, u8, u8)> {
-    bail!("not implemented");
+    Err(anyhow!("not implemented"))
 }
 
 /// Returns the current time since epoch in nanoseconds.
-#[tracing::instrument(name = "time.now_ns", err)]
-pub fn now_ns() -> Result<i64> {
-    bail!("not implemented");
+#[tracing::instrument(name = "time.now_ns")]
+pub fn now_ns() -> i64 {
+    let utc: DateTime<Utc> = Utc::now();
+    utc.timestamp_nanos()
 }
 
 /// Returns the duration in nanoseconds represented by a string.
 #[tracing::instrument(name = "time.parse_duration_ns", err)]
-pub fn parse_duration_ns(duration: String) -> Result<i64> {
-    bail!("not implemented");
+pub fn parse_duration_ns(duration: String) -> Result<u128> {
+    Ok(duration_str::parse(duration.as_str())?.as_nanos())
 }
 
 /// Returns the time in nanoseconds parsed from the string in the given format.
 /// `undefined` if the result would be outside the valid time range that can fit
 /// within an `int64`.
 #[tracing::instrument(name = "time.parse_ns", err)]
 pub fn parse_ns(layout: String, value: String) -> Result<i64> {
-    bail!("not implemented");
+    Err(anyhow!("not implemented"))
 }
 
 /// Returns the time in nanoseconds parsed from the string in RFC3339 format.
 /// `undefined` if the result would be outside the valid time range that can fit
 /// within an `int64`.
 #[tracing::instrument(name = "time.parse_rfc3339_ns", err)]
 pub fn parse_rfc3339_ns(value: String) -> Result<i64> {
-    bail!("not implemented");
+    Ok(DateTime::parse_from_rfc3339(value.as_ref())?.timestamp_nanos())
 }
 
 /// Returns the day of the week (Monday, Tuesday, ...) for the nanoseconds since
 /// epoch.
 #[tracing::instrument(name = "time.weekday", err)]
 pub fn weekday(x: serde_json::Value) -> Result<String> {
-    bail!("not implemented");
+    let date_time = extract_data_from_value(&x)?;
+
+    Ok(match date_time.weekday() {
+        Weekday::Mon => "Monday",
+        Weekday::Tue => "Tuesday",
+        Weekday::Wed => "Wednesday",
+        Weekday::Thu => "Thursday",
+        Weekday::Fri => "Friday",
+        Weekday::Sat => "Saturday",
+        Weekday::Sun => "Sunday",
+    }
+    .to_string())
+}
+
+/// convert nanoseconds to [`chrono::DateTime`] by the given timezone
+///
+/// # Errors
+///
+/// - the `time_zone` when could not parse the given timezone
+/// - the given `ns` representation is invalid.
+fn nanoseconds_to_date(ns: i64, time_zone: &str) -> Result<DateTime<Tz>> {
```

Comment:
Here I'd make `time_zone` an `Option<&str>` and do something like

```rust
let tz = time_zone
  .map(|tz| tz.parse())
  .transpose()?
  .unwrap_or(Tz::UTC);
```


which avoids having to parse `"UTC"` or having to define a `DEFAULT_TIMEZONE` const